File: 第七次课/第二题/test2.py
# -*- coding: utf-8 -*-
import numpy as np


def xk(x,L):
    y=[]
    y.append(x)
    for n in range(1,L):
        p=x*y[-1]
        y.append(p)
        
    return y






N=2000
x=np.sort(np.random.random(N)*2-1)       #[-1,1]随机生成N个点，从小到大排序

y=np.zeros(N)
for k in range(0,N):
    y[k]=1/(1+25*x[k]**2)                     
    y[k]=y[k]+(np.random.random()*2-1)*0.1         #在y上加入扰动



#构建数据集：

L=25         #用25次多项式进行逼近
data=[]

for k in range(0,N):
    data.append(xk(x[k],L))

data=np.array(data)

label=np.array([y]).T

''''' LinearRegression '''  
X=np.hstack((data,np.ones(label.shape)))
Y=label

W=np.dot(np.linalg.pinv(X),Y)#利用广义逆求系数
w=W[0:-1]
b=W[-1]

y_train_predict=np.dot(data,w)+b
''''' 线性回归建模 '''  




''' 测试模型 '''
M=120
x_test=np.sort(np.random.random(M)*2-1)       #[-1,1]随机生成M个点，从小到大排序

y_test=np.zeros([M,1])
for k in range(0,M):
    y_test[k]=1/(1+25*x_test[k]**2)

    
data_test=[]
for k in range(0,M):
    data_test.append(xk(x_test[k],L))

data_test=np.array(data_test)

y_predict=np.dot(data_test,w)+b



''' 测试集误差 '''

Error_test=np.sqrt(np.sum((y_predict-y_test)**2)/len(y_test))
print('Error_test',Error_test)




''' 画图 '''

import matplotlib.pyplot as plt  

''' 训练集 '''
fig1 = plt.figure('fig1')
plt.plot(x,y,'r+',lw=1)    #训练集数据
plt.plot(x,y_train_predict,'b',lw=1)    #模型


''' 测试集 '''
fig2 = plt.figure('fig2')
plt.plot(x_test,y_predict,'b',lw=1)   #预测值
plt.plot(x_test,y_test,'r+',lw=1)     #测试集精确值

File: 第七次课/第二题/test2.m
clear

N=2000;


%generate random points in (-1,)]
xi=sort(rand(1,N)*2-1);
yi=1./(1+25*(xi.^2))+(rand(1,N)*2-1)*0.1;
% scatter(xi,yi,'red');hold on;


% generate X
L=25;
start=1;
X=zeros(N,L+1);
% X(:,1)=1;%x^0=1;
for i = 1:N
    for j =start:L
        X(i,j+1)=xi(i)^j;
    end
end


%calculate the A
Y=yi';
A=pinv(X'*X)*X'*Y;
%A=X\Y;
% scatter(xi,(X*A)','blue');

y_train_predict=(X*A)';

%generate xj
M=120;
x=sort(rand(1,M)*2-1);


%generate X_pre
X_pre=zeros(M,L+1);
for i = 1:M
    for j =start:L
        X_pre(i,j+1)=x(i)^j;
    end
end


%calculate y_pre
y_pre=(X_pre*A)';
%error
y_exact=1./(1+25*x.^2);
%average error
error=sqrt(sum((y_pre-y_exact).^2)/length(y_exact))

%训练集
subplot(2,1,1)
scatter(xi,yi,'blue');hold on
plot(xi,y_train_predict,'red');

%测试集
subplot(2,1,2)
plot(x,y_pre,'blue');hold on
scatter(x,y_exact,'red');
